File: Tree/114FlattenBinaryTreetoLinkedList.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        helper(root, NULL);
    }
    TreeNode* helper(TreeNode* root, TreeNode* pre){
        if(!root) return pre;
        pre = helper(root->right, pre);
        pre = helper(root->left, pre);
        root->right = pre;
        root->left = NULL;
        pre = root;
        return pre;
    }
};

File: Tree/114FlattenBinaryTreetoLinkedList.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: None Do not return anything, modify root in-place instead.
        """
        prev = None
        self.helper(root, prev)
        
    def helper(self, root, prev):
        if not root:
            return prev
        prev = self.helper(root.right, prev)
        prev = self.helper(root.left, prev)
        
        root.right = prev
        root.left = None
        prev = root
        return prev